### PR TITLE
fix: Websocket Reconnection on React Native

### LINKED ISSUE #5998: bug: Websocket does not gracefully resume/reconnect on iOS
### Provide environment information

```
 System:
    OS: Linux 6.9 Pop!_OS 22.04 LTS
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
    Memory: 8.56 GB / 15.26 GB
    Container: Yes
    Shell: 5.8.1 - /usr/bin/zsh
  Binaries:
    Node: 22.8.0 - ~/.volta/tools/image/node/22.8.0/bin/node
    Yarn: 1.22.19 - ~/.volta/tools/image/yarn/1.22.19/bin/yarn
    npm: 10.8.2 - ~/.volta/tools/image/node/22.8.0/bin/npm
    pnpm: 9.9.0 - ~/.local/share/pnpm/pnpm
    bun: 1.1.28 - ~/.bun/bin/bun
  npmPackages:
    @tanstack/react-query: ^5.56.2 => 5.56.2 
    @trpc/client: ^11.0.0-rc.502 => 11.0.0-rc.502+2a8c56027 
    @trpc/react-query: ^11.0.0-rc.502 => 11.0.0-rc.502+2a8c56027 
    @trpc/server: ^11.0.0-rc.502 => 11.0.0-rc.502+2a8c56027 
    react: 18.2.0 => 18.2.0 
    typescript: ^5.6.2 => 5.6.2 
 ```

### Describe the bug

I think I am seeing two bugs with `wsLink` specific on iOS on React Native

- Active subscriptions do not resume when woken up from being locked 😴 
- If `lazy` is enabled, an iOS device will not reconnect to the websocket server at all when woken up from being locked 😴 

I get this error in the wsLink's `onError` whenever I wakeup my iOS device form being locked

### Link to reproduction

https://github.com/bnussman/trpc-ws-issue

### To reproduce

1. Downlaod [Expo Go](https://apps.apple.com/us/app/expo-go/id982107779) on your iOS device
2.  Clone the reproduction repository 
3.  Install dependencies with
    ```
    bun install
    ```
4.  Start the dev servers with 
    ```
    bun start
    ```
5. On your iOS device, open Expo Go and open this project
6. Observe the counter is counting up (meaning the subscription is working and connected)
7. Lock your iOS device's screen and wait for the server to register that the device disconnected
8. Unlock your iOS device and observe that the WS connections, but the tRPC subscription does not resume
9. Repeat steps 5-8 with lazy mode enabled. Observe that when your iOS device is woken up from sleep, the WS does not reconnect at all

### Additional information

This also happens in my production iOS app, so I have confirmed that this is not specific to Expo Go. 

It might be a React Native bug, but I haven't found anything documenting if this is the case. It maybe me something tRPC needs to address for compatibility with React Native

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5998">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5998/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by bnussman**
Closes #5998

## 🎯 Changes

- Makes a small change in the websocket client that always sets the connection state to `connecting` when `tryReconnect` is called so that `pendingRequests` can not be deleted when the connection is closed for an unexpected reason

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

**Review by KATT — COMMENTED** [review]
I'm pretty sure this PR as-is doesn't impact the problem outlined, I don't see how it makes anything really, if there is a connectTimer it will eventually reconnect anyway

**Comment by bnussman** [comment]
Yeah, I wish I could justify why/how this work, but it does do the job for whatever reason.

Whenever I unlock my phone I the connection state is `closed` and this code runs:

https://github.com/trpc/trpc/blob/1d65ff96f26214f80703b95a937171aabb385dd6/packages/client/src/links/wsLink.ts#L384-L389

Which ends up incorrectly deleting all of the pending subscriptions.

I can't really figure out why the state never becomes `connecting`

**Comment by KATT** [comment]
Yeah that file is a bit spaghetti, hard to follow anything that goes on in there, missing some sort of state machine.

Will properly review this when I get the time to think if it has any side effects

A regression test that proves this fixes something would be useful, might break in the future otherwise

File: packages/client/src/links/wsLink.ts [hunk 1]
```diff
@@ -369,38 +398,8 @@ export function createWSClient(opts: WebSocketClientOptions) {
...
-          if (req.connection !== self) {
-            continue;
-          }
-
-          if (self.state === 'closed') {
-            // If the connection was closed, we just call `complete()` on the request
-            delete pendingRequests[key];
-            req.callbacks.complete?.();
-            continue;
-          }
```
Comment (KATT) [inline packages/client/src/links/wsLink.ts:389]: this part was simply removed

**Review by KATT — COMMENTED** [review]
cc @bnussman could you test with the changes i pushed?

`pendingRequests` represents any request that has been sent but not gotten a response yet, so they should be aborted _except_ subscriptions which should gracefully resume

there seems to have been a bug where we deleted subscriptions in some cases whereas they should've been gracefully resumed instead

**Comment by KATT** [comment]
cc @bnussman I'm going to merge this as all the tests are passing, please re-open if it doesn't solve it

**Review by KATT — APPROVED** [review]

**Comment by bnussman** [comment]
@KATT  

Sorry for being slow to test over the weekend. I've been out of town!

Just tested `11.0.0-rc.528` and I can confirm Websocket reconnection works perfect for me now! Huge thank you ❤️

**Comment by bnussman** [comment]
Did a one-time sponsor for your time! Thanks again 💵

**Comment by KATT** [comment]
> Did a one-time sponsor for your time! Thanks again 💵

Appreciate it! 🙏
